### PR TITLE
Added fluent API interface to SPI API class.

### PR DESCRIPTION
**Pull request by bremoran**
Convert many of the transfer management functions to operate on transaction_data_t instead of parameters.

@autopulated @hugovincent @bogdanm @0xc0170

### CODE REVIEW

File: test/asynch_spi/asynch_spi.cpp [hunk 1]
```diff
@@ -58,7 +58,12 @@ class SPITest {
         printf("Starting short transfer test\r\n");
         init_rx_buffer();
         cs = 0;
-        printf("Res is %d\r\n", spi.transfer(tx_buf, SHORT_XFR, rx_buf, SHORT_XFR, SPI::event_callback_t(this, &SPITest::short_transfer_complete_cb), SPI_EVENT_COMPLETE));
+        int rc = spi.transfer()
+            .tx(tx_buf, SHORT_XFR)
+            .rx(rx_buf, SHORT_XFR)
+            .callback(SPI::event_callback_t(this, &SPITest::short_transfer_complete_cb), SPI_EVENT_COMPLETE)
+            .apply();
```
Comment (bogdanm) [inline test/asynch_spi/asynch_spi.cpp]: Isn't `apply` redundant here ? I thought we are doing 'execute on destruction'.

Comment (bremoran) [inline test/asynch_spi/asynch_spi.cpp]: ```apply()``` allows you to reap status output.  The other option is to provide a cast-to-int.

**Comment by autopulated** [comment]
I think it'd be good to add at least asserts preventing .tx(...).tx(...) and other invalid combinations before merging.

File: mbed/SPI.h [hunk 2]
```diff
@@ -101,41 +100,70 @@ class SPI {
...
+        /** Set the receive buffer
+         *  Sets the receive buffer pointer and receive size
+         *
+         *  NOTE: Repeated calls to rx() override buffer parameters.
+         *
+         *  @param[in] rxBuf a pointer to the receive buffer
+         *  @param[in] rxSize the size of the receive buffer
+         *  @return a reference to the SPITransferAdder
+         */
+        SPITransferAdder & rx(void * rxBuf, size_t rxSize);
```
Comment (0xc0170) [inline mbed/SPI.h]: Can you please format the new code to be consistent with the current implementation?
For note we can use '@note'

**Comment by bremoran** [comment]
@autopulated @0xc0170 @bogdanm, are you happy with this PR as it is now?

**Comment by autopulated** [comment]
LGTM

**Comment by bremoran** [comment]
Looks like I'll have to rebase first.

**Comment by 0xc0170** [comment]
LGTM, as this is breaking change, I would like at least one more +1.. We got already on master some changes for new minor.

**Comment by bogdanm** [comment]
Looks good! Let's give it a go. Very nice work @bremoran .

**Pull request merged**